contracts/SafeProtocolMediator.sol

```diff
@@ -0,0 +1,241 @@
+    ) external override onlyEnabledModule(safe) returns (bytes[] memory data) {
+        data = new bytes[](transaction.actions.length);
+        for (uint256 i = 0; i < transaction.actions.length; ++i) {
+            SafeProtocolAction memory safeProtocolAction = transaction.actions[i];
+            (bool isActionSuccessful, bytes memory resultData) = safe.execTransactionFromModuleReturnData(
+                safeProtocolAction.to,
+                safeProtocolAction.value,
+                safeProtocolAction.data,
+                0
+            );
+
+            // Need to revisit the approach below. If some actions fail, the transaction stiil succeeds.
+            // With current approach, even if one action fails, `data` will be empty bytes even for successful
+            // actions.
+            if (!isActionSuccessful) {
+                revert ActionExecutionFailed(address(safe), transaction.metaHash, i);
+            } else {
+                data[i] = resultData;
+            }
+        }
+
+        emit ActionsExecuted(address(safe), transaction.metaHash, transaction.nonce);
+    }
+
+    /**
+     * @notice This function executes a delegate call on a safe if the module is enabled and
+     *         root access it granted.
+     * @param safe A Safe instance
+     * @param rootAccess A struct of type SafeRootAccess containing information of about the action to be executed.
+     *                   Users can add logic to validate metahash through a transaction guard.
+     * @return data bytes types containing the result of the executed action.
+     */
+    function executeRootAccess(
+        ISafe safe,
+        SafeRootAccess calldata rootAccess
+    ) external override onlyEnabledModule(safe) returns (bytes memory data) {
+        SafeProtocolAction memory safeProtocolAction = rootAccess.action;
+
+        if (!ISafeProtocolModule(msg.sender).requiresRootAccess() || !enabledModules[address(safe)][msg.sender].rootAddressGranted) {
+            revert ModuleRequiresRootAccess(msg.sender);
+        }
+
+        bool success;
+        (success, data) = safe.execTransactionFromModuleReturnData(
+            safeProtocolAction.to,
+            safeProtocolAction.value,
+            safeProtocolAction.data,
+            1
+        );
+        if (success) {
+            emit RootAccessActionExecuted(address(safe), rootAccess.metaHash);
+        } else {
+            revert RootAccessActionExecutionFailed(address(safe), rootAccess.metaHash);
+        }
+    }
+
+    /**
+     * @notice Called by a Safe to enable a module on a Safe. To be called by a safe.
+     * @param module ISafeProtocolModule A module that has to be enabled
+     * @param allowRootAccess Bool indicating whether root access to be allowed.
+     */
+    function enableModule(address module, bool allowRootAccess) external noZeroOrSentinelModule(module) {
+        if (enabledModules[msg.sender][module].nextModulePointer != address(0)) {
```

Comment:
You can save some gas by caching msg.sender into a variable

Reply:
Using `msg.sender` is cheaper than using local variable.
I tried in the commit here: https://github.com/5afe/safe-protocol/commit/c441e0cefc3ebd52f621f792f2dc9213f4e21857

Using `msg.sender` (no local variable)

<img width="871" alt="Screenshot 2023-07-02 at 12 36 25" src="https://github.com/5afe/safe-protocol/assets/16358982/3aaf468d-1040-41b9-9f29-ea15e42e1839">

Using local variable

<img width="1066" alt="Screenshot 2023-07-02 at 12 39 07" src="https://github.com/5afe/safe-protocol/assets/16358982/472df0c4-7fef-4d75-a846-8c317ded552d">

Also other ref: https://stackoverflow.com/questions/64854256/how-much-accessing-msg-sender-costs-is-it-useful-to-store-it-in-a-variable-and

Reply:
Oh, interesting, that make sense. Although I'm not sure if the addresses get stored in memory, I think they remain on the stack.

I took another look at the function and I though it was possible to shave off some gas on the SLOADs, here is my attempt:
```
    function enableModule(address module, bool allowRootAccess) external noZeroOrSentinelModule(module) {
        ModuleAccessInfo storage senderSentinelModule = enabledModules[msg.sender][SENTINEL_MODULES];
        ModuleAccessInfo storage senderModule = enabledModules[msg.sender][module];

        if (senderModule.nextModulePointer != address(0)) {
            revert ModuleAlreadyEnabled(msg.sender, module);
        }

        bool requiresRootAccess = ISafeProtocolModule(module).requiresRootAccess();
        if (allowRootAccess != requiresRootAccess) {
            revert ModuleAccessMismatch(module, requiresRootAccess, allowRootAccess);
        }

        if (senderSentinelModule.nextModulePointer == address(0)) {
            senderSentinelModule.rootAddressGranted = false;
            senderSentinelModule.nextModulePointer = SENTINEL_MODULES;
        }

        senderModule.nextModulePointer = senderSentinelModule.nextModulePointer;
        senderModule.rootAddressGranted = allowRootAccess;
        senderSentinelModule.nextModulePointer = module;

        emit ModuleEnabled(msg.sender, module, allowRootAccess);
    }
    ```
    
This uses 76237 gas for me instead of 77507

Reply:
https://github.com/5afe/safe-protocol/commit/54d6117839548907243542f98ec683060fb5e168